File: server/app/schemas/messages.py
"""WebSocket message schema, as specified in docs/architecture/phase1-design.md.

Provider identity (STT/TTS/LLM vendor) is intentionally absent from every payload here:
the agent's behavior depends only on message content, not on which provider produced or
will consume it.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


# ---- shared value types -----------------------------------------------------

class Selection(BaseModel):
    start_line: int
    end_line: int


class Cursor(BaseModel):
    line: int
    col: int


class EditorContext(BaseModel):
    file_path: str
    language: str
    selection: Selection | None = None
    cursor: Cursor | None = None
    file_text: str


class Diagnostic(BaseModel):
    severity: Literal["error", "warning", "info", "hint"]
    line: int
    message: str


# ---- client -> server ---------------------------------------------------------

class VoiceUtterancePayload(BaseModel):
    text: str
    is_final: bool
    context: EditorContext


class VoiceUtterance(BaseModel):
    type: Literal["voice.utterance"] = "voice.utterance"
    id: str
    payload: VoiceUtterancePayload


class DiffDecisionPayload(BaseModel):
    proposal_id: str
    decision: Literal["accepted", "rejected"]


class DiffDecision(BaseModel):
    type: Literal["diff.decision"] = "diff.decision"
    id: str
    payload: DiffDecisionPayload


class SessionDiagnosticsPayload(BaseModel):
    file_path: str
    diagnostics: list[Diagnostic]


class SessionDiagnostics(BaseModel):
    type: Literal["session.diagnostics"] = "session.diagnostics"
    payload: SessionDiagnosticsPayload


# ---- server -> client ---------------------------------------------------------

class AgentExplanationDeltaPayload(BaseModel):
    text: str
    done: bool


class AgentExplanationDelta(BaseModel):
    type: Literal["agent.explanation.delta"] = "agent.explanation.delta"
    id: str
    payload: AgentExplanationDeltaPayload


class DiffProposedPayload(BaseModel):
    proposal_id: str
    file_path: str
    unified_diff: str
    rationale: str


class DiffProposed(BaseModel):
    type: Literal["diff.proposed"] = "diff.proposed"
    id: str
    payload: DiffProposedPayload


class ErrorPayload(BaseModel):
    code: str
    message: str


class ErrorMessage(BaseModel):
    type: Literal["error"] = "error"
    id: str | None = None
    payload: ErrorPayload

File: client/src/connection/messages.ts
/**
 * WebSocket message schema, mirrored from server/app/schemas/messages.py.
 * Keep both in sync manually for Phase 1 -- see docs/architecture/phase1-design.md.
 *
 * Provider identity (STT/TTS/LLM vendor) is intentionally absent from every payload:
 * the agent's behavior depends only on message content, not on which provider produced
 * or will consume it.
 */

export interface Selection {
  startLine: number;
  endLine: number;
}

export interface Cursor {
  line: number;
  col: number;
}

export interface EditorContext {
  filePath: string;
  language: string;
  selection?: Selection;
  cursor?: Cursor;
  fileText: string;
}

export interface Diagnostic {
  severity: "error" | "warning" | "info" | "hint";
  line: number;
  message: string;
}

// ---- client -> server ----

export interface VoiceUtterance {
  type: "voice.utterance";
  id: string;
  payload: {
    text: string;
    isFinal: boolean;
    context: EditorContext;
  };
}

export interface DiffDecision {
  type: "diff.decision";
  id: string;
  payload: {
    proposalId: string;
    decision: "accepted" | "rejected";
  };
}

export interface SessionDiagnostics {
  type: "session.diagnostics";
  payload: {
    filePath: string;
    diagnostics: Diagnostic[];
  };
}

export type ClientToServerMessage = VoiceUtterance | DiffDecision | SessionDiagnostics;

// ---- server -> client ----

export interface AgentExplanationDelta {
  type: "agent.explanation.delta";
  id: string;
  payload: {
    text: string;
    done: boolean;
  };
}

export interface DiffProposed {
  type: "diff.proposed";
  id: string;
  payload: {
    proposalId: string;
    filePath: string;
    unifiedDiff: string;
    rationale: string;
  };
}

export interface ErrorMessage {
  type: "error";
  id?: string;
  payload: {
    code: string;
    message: string;
  };
}

export type ServerToClientMessage = AgentExplanationDelta | DiffProposed | ErrorMessage;
